Realm/RLMRealm.h

```diff
@@ -500,6 +505,96 @@ typedef void (^RLMNotificationBlock)(RLMNotification notification, RLMRealm *rea
  */
 - (BOOL)transactionWithoutNotifying:(NSArray<RLMNotificationToken *> *)tokens block:(__attribute__((noescape)) void(^)(void))block error:(NSError **)error;
 
+#ifdef REALM_ASYNC_WRITES
+
+/**
+ Indicates if the Realm is currently engaged in an async write transaction.
+ 
+ @warning Do not simply check this property and then start a write transaction whenever an object needs
+          to be created, updated, or removed. Doing so might cause a large number of write transactions
+          to be created, degrading performance. Instead, always prefer performing multiple updates
+          during a single transaction.
+ */
+@property (nonatomic, readonly) BOOL inAsyncWriteTransaction;
+
+/// A block type representing a block which can be used to report an async transaction related error to the application.
+typedef void(^RLMRealmAsyncErrorHandler)(RLMAsyncTransactionId, NSError *);
+
+/**
+ Set the error handler for the asynbhronous transactions.
+ Synchronous try/catch hadling will not work with the asynchronous transactions.
+ Asynchronous exceptions will be dispatched to the handler as `NSErrors`.
+ @note pass `nil` as a parameter to remove the handler
+ */
+- (void)setAsyncErrorHandler:(nullable RLMRealmAsyncErrorHandler)block;
+
+/**
+ Begins asynchronous write transaction.
+ 
+ @param block The block containing actions to perform.
+             `block` should end by calling `commitAsyncWriteTransaction`, `cancelAsyncTransaction:`,
+             `commitWriteTransaction` or `cancelWriteTransaction`.
+             Returning without one of these calls will be equivalent to calling `cancelAsyncTransaction`.
+ 
+ @return Asynchronous transaction's Id.
+ 
+ @note `block` is queued for execution on the scheduler associated with the current realm.
+       It will run after the write mutex has been acquired.
+       The call returns immediately allowing the caller to proceed while the write mutex is held by someone else.
+       Write blocks from multiple calls to `beginAsyncWriteTransaction:` or `asyncTransactionWithBlock:`
+       will be executed in order.
+       A later call to `beginAsyncWriteTransaction:` or `asyncTransactionWithBlock:` will wait for any earlier
+       write blocks.
+ */
+- (RLMAsyncTransactionId)beginAsyncWriteTransaction:(void(^)())block;
+
+/** Commit asynchronous transaction.
+ 
+ @param doneBlock  is queued for execution on the scheduler associated with
+                  the current realm. It will run after the commit has reached stable storage.
+ 
+ @param isGroupingAllowed If `true`, the next `commitAsyncWriteTransaction` *may* run without an
+                         intervening synchronization of stable storage.  Such a sequence of commits
+                         form a group. In case of a platform crash, either none or all of the commits
+                         in a group will reach stable storage.
```

Comment:
I may have missed a larger design discussion - but would this violate atomicity?
I suppose since this is opt-in, and the developer is warned it's fine?

Reply:
I add  "false by default" to make it obvious.